### ISSUE TITLE
Bug when serializing java.util.Date

### ISSUE THREAD

**Issue by vic-cw** [konsoletyper/teavm-flavour#24]
I get a bug when trying to launch an AJAX request through the `RESTClient.factory(...).createResource(...)` way, when data to serialize contains a java.util.Date.

My guess is that it is because the `time` field of the underlying `Calendar` object has not been initialized, and thus in Javascript it is undefined whereas in Java it is 0.

Stack trace in browser:

> runtime.js:664 Uncaught TypeError: Cannot read property 'hi' of undefined
>     at Long_eq (runtime.js:664)
>     at ju_Calendar_setTimeInMillis (TCalendar.java:455)
>     at ju_Calendar_setTime (TCalendar.java:451)
>     at jt_SimpleDateFormat_format (TSimpleDateFormat.java:57)
>     at jt_DateFormat_format (TDateFormat.java:85)
>     at otfjs_JsonSerializer$proxy0_serialize (JsonSerializerEmitter.java:467)
>     at otfjs_JsonSerializer$proxy0.main [as $serialize1] (TCollections.java:547)
>     at otfj_JSON_serialize1 (JSON.java:52)
>     at otfj_JSON_serialize (JSON.java:41)

My interpretation of where it fails:

- at line 1797 of `Calendar.class`, which is `time == millis`
- which is called by `Calendar.setTime` on line 943 of `SimpleDateFormat`
- which is the main `format` method of `SimpleDateFormat`

**Comment by vic-cw** [comment]
Actually I found out that the problem I ran into was not due to a bad serialization of Dates.

What happened was that the objects that I thought were Dates were actually Strings. This is because I had used a `html:bidir-value` on an `input` with `type="date"` and I naively thought that this would translate the inputs into Date objects, whereas they ended up being Strings, thus the failure.

I guess the solution then would be to fail at compile time if there is a bidir-value bound to a non String object.

**Comment by ScraM-Team** [comment]
Good find, thanks.  The built-in Flavour components use the standard, documented API, so I will take a look and see if I can see where this happens.

**Comment by ScraM-Team** [comment]
https://github.com/konsoletyper/teavm-flavour/blob/master/templates/src/main/java/org/teavm/flavour/components/html/BidirectionalValueBinder.java#L56

shows that it is converting the value to a string before setting it into the input.  And it looks like it receives the value from the input as a string as well.

To do your own conversion, you can use html:bidir-value="myfield".  Then implement getters and setters for myfield.  Although they should take and return Strings, inside the method you can convert to a Date and store that in your Date field.

**Comment by vic-cw** [comment]
@ScraM-Team thanks for the swift reaction.

Yes indeed this is what I end up doing.

However, to avoid other people losing time figuring out why it fails, I think it would be beneficial to detect this wrong binding at compile time.

I can try to submit a pull request in some time.

**Comment by ScraM-Team** [comment]
I agree, a nice warning message at compile time would save people some trouble.  That would be great.

**Comment by vic-cw** [comment]
Cool, I'll try to do that as soon as possible

**Comment by konsoletyper** [comment]
Instead of `<input type="date"` you can use Flavour calendar component: http://teavm.org/docs/flavour/advanced-ui-components.html

**Comment by vic-cw** [comment]
@konsoletyper thank! I had seen it but not tried it yet. I'll check it out

**Comment by vic-cw** [comment]
@ScraM-Team I would love to have your opinion on [this](https://github.com/renatoathaydes/teavm-gradle-plugin/pull/4) if you have some time!
